set.mm

```diff
@@ -634631,6 +634631,42 @@ deduction version (Contributed by metakunt, 22-May-2024.) $)
       DVEVFVGWEWRWSVIWMWEVQEWDWOAWQVEAWGWDWHVJVKVJVLUHVMVNVNIVO $.
   $}
 
+  ${
+    $d I x $.  $d M x $.  $d X x $.  $d ph x $.
+    metakunt3.1 $e |- ( ph -> M e. NN ) $.
+    metakunt3.2 $e |- ( ph -> I e. NN ) $.
+    metakunt3.3 $e |- ( ph -> I <_ M ) $.
+    metakunt3.4 $e |- A = ( x e. ( 1 ... M ) |-> if (
+       x = I , M , if ( x < I , x , ( x - 1 ) ) ) ) $.
+    metakunt3.5 $e |- ( ph -> X e. ( 1 ... M ) ) $.
+    $( Value of A. (Contributed by metakunt, 23-May-2024.) $)
+    metakunt3 $p |- ( ph -> ( A ` X ) = if (
+       X = I , M , if ( X < I , X , ( X - 1 ) ) ) ) $=
```

Comment:
Depending on your goals, it may be worth parameterizing this A by I:
```
A = ( i e. ( 1 ... M ) |-> ( x e. ( 1 ... M ) |-> if ( .... ) )
```
and similarly for the A defined in the next statement.

Reply:
Thank you for your feedback.

Actually, that is an excellent point. I came a long way since I started with metamath a few weeks ago, see https://github.com/metamath/set.mm/issues/3908 what I want to prove. Back then the theorem I want to prove now was way too difficult for me. But now once I have proven the lcm inequality lemma I am confident to once again tackle it.

A is actually parametrised by I, but what I first want to show is that for any I the function is a permutation. I don't see the benefits in parametrising it right now, as the theorems that I need in this stage would be more complex to prove. Are there any drawbacks in keeping it as is or can I later parametrise it if I choose so? This is a sincere question since I am not that experienced in metamath yet and I want to avoid potential pitfalls.

I have also thought about defining it via operation as you could write something like ( I A X ) and then evaluate it.
One theorem would then be ( X A X ) = X for all X in 1 ... M, which is later an important invariant that holds. But so far this just seems syntactic sugar which I am not yet sure if I'll need. I also think that It should be possible to do when needed without much trouble.

My goal is also mentioned in https://github.com/metamath/set.mm/issues/3990 I want to use https://us.metamath.org/mpeuni/2fvidf1od.html and for that both functions defined need to be maps from 1,...m to itself.

Reply:
You can probably add the functionalized version later, and then a proof for the new `` ( YourFunction ` i ) `` will be able to reuse the result that you just proved in this PR since it satisfies its hypotesis. And defining the parameterized function will certainly not spare you that work anyway. So maybe it is a bit premature indeed.

And actually, your function could have two parameters, e.g., $F_n^i \in \mathfrak{S}_n$ with $F(i) = n$ and the other values as you wrote. This is how I view it from a mathematical point of view without looking at formalization matters. Maybe it is overkill for the formalization work, and only makes it harder for no immediate benefit. The benefit really depends on what you will do with it later.  If $F$ were really a fundamental object, I would push towards the parameterized version, but given that it's not, maybe it's not worth the hassle.

Indeed, for binary functions, it may be better to use the notation `( x e. A , y e. B |-> ... )` and then use the operation notation `( X YourFunction Y )` like you wrote. It depends on the manipulation lemmas available in set.mm, and I don't know exactly the current situation.

Reply:
> Indeed, for binary functions, it may be better to use the notation `( x e. A , y e. B |-> ... )` and then use the operation notation `( X YourFunction Y )` like you wrote. It depends on the manipulation lemmas available in set.mm, and I don't know exactly the current situation.

Yes, this would be the more preferable way.

Reply:
Thanks for your feedback.

As of now the lemmas are transitionary. What I want is to show existence of a map
```math
f: \{1,\dots,m\} \to \ ( C_{2} \times S_{n})
```
```math
i \mapsto (c_i,d_i)
```
where $d_i(i)=i$ and some other conditions. This lemma constructs this map and shows that this condition and other conditions also hold.
You can think of $d_i = A^{-1}_i\circ B_i \circ A_i$ and so far the I want to show that those maps are inverse of each other and thus are both permutations. Next step is to construct B with the same procedure. Then I show that $d_i$ is a permutation and thus I have constructed $d_i$. Next I show $d_i(i)=i$ and finally I construct f.
But so far I am not there yet. There are still also some open points as I will need a Homomorphism from $\mathbb{Z} \to C_2$ where $C_2$ is identified by $\mathbb{Z}/{2\mathbb{Z}}$

I don't know what lemmas we have for finite fields, as one map is $e: S_n \to C_2$, plugging the constructed $f$ into e should yield $e(f)\equiv 0$